sstables: make_entry_descriptor: make regex non-greedy

bhalevy opened the pull request:
With greedy matching, an sstable path in a snapshot directory with a tag that ressembles a name-<uuid> would match the dir regular expression as the longest match, while a non-greedy regular expression would correctly match the real keyspace and table as the shortest match.

Also, add a regression unit test reproducing the issue and validating the fix.

Fixes #25242

* The issue exists forever, but was produced only recently with https://github.com/scylladb/scylla-ccm/pull/676. For the sake of being conservative with backports, no backport requested at the moment

bhalevy commented:
@denesb the test tripped on the following ASAN issue:
```
DEBUG 2025-08-05 03:46:51,543 [shard 0:sl:d] scylla-sstable - query_operation(): running query SELECT * FROM scylla_sstable.cql_test_1754354809465 
AddressSanitizer:DEADLYSIGNAL
=================================================================
==378734==ERROR: AddressSanitizer: SEGV on unknown address (pc 0x00000416e5ca bp 0x7ffd56d37330 sp 0x7ffd56d370f0 T0)
==378734==The signal is caused by a READ memory access.
==378734==Hint: this fault was caused by a dereference of a high value address (see register values below).  Disassemble the provided pc to learn which register was used.
    #0 0x00000416e5ca in managed_bytes_basic_view<(mutable_view)0>::managed_bytes_basic_view(managed_bytes const&) ./utils/managed_bytes.hh:396:59
    #1 0x00000b2c46b5 in (anonymous namespace)::print_query_results_json(cql3::result const&) tools/scylla-sstable.cc:2860:74
    #2 0x00000b2c46b5 in (anonymous namespace)::query_operation_result_visitor::visit(cql_transport::messages::result_message::rows const&) tools/scylla-sstable.cc:2891:17
    #3 0x00000b2ec70d in (anonymous namespace)::query_operation(seastar::lw_shared_ptr<schema const>, reader_permit, std::vector<seastar::lw_shared_ptr<sstables::sstable>, std::allocator<seastar::lw_shared_ptr<sstables::sstable>>> const&, sstables::sstables_manager&, boost::program_options::variables_map const&)::$_0::operator()(cql_test_env&) (.resume) tools/scylla-sstable.cc:3046:17
```

Care to look into it?

denesb commented on test/cqlpy/test_tools.py:
```diff
@@ -1731,3 +1731,25 @@ def test_scylla_sstable_query_temp_dir(cql, scylla_path, scylla_data_dir):
+def test_scylla_sstable_dump_data_from_snapshot(cql, test_keyspace, scylla_path, scylla_data_dir, tag):
+    with scylla_sstable(simple_clustering_table, cql, test_keyspace, scylla_data_dir) as (table, schema_file, sstables):
+        host = cql.cluster.contact_points[0]
+        args = [scylla_path, "nodetool", "snapshot", "--host", host, "--tag", tag, "--keyspaces", test_keyspace]
```
We have cqlpy/nodetool.py which abstracts which nodetool we use.
For scylla-only tests such as this, I think using scylla-nodetool direclty is fine.

nyh replied:
I agree t's not a big problem, especially since the whole point of this test file (test_tools.py) is to exercise various scylla "tools", so how can we complain when another one (scylla nodetool) is being exercised? :-)

But the other view is that this test function (not the file) was supposed to be about "scylla sstable", not about "scylla nodetool", so it's better not to mix it into the same test. We want a bug in scylla nodetool to trip tests for scylla nodetools - not tests for scylla sstable. As @denesb rightly noted we have (from cqlpy/nodetool.py)  `nodetool.take_snapshot()` which you can use and doesn't rely on "scylla nodetool" (and could have worked on Cassandra too, but in this specific test it's not interesting).

Last but not least, the nodetool.py function is faster to call than to start a separate executable (perhaps not much, though, I didn't measure).

nyh replied:
But please seem my rebuttal of this argument :-) I'm not saying it's a big problem, but using `nodetool.take_snapshot()` and `nodetool.del_snapshot()` is better and faster so why not?

bhalevy replied:
> But please see my rebuttal of this argument, above :-) I'm not saying it's a big problem, but using `nodetool.take_snapshot()` and `nodetool.del_snapshot()` is better and faster so why not?

okay, fair enough

nyh commented on test/cqlpy/test_tools.py:
```diff
@@ -1731,3 +1731,25 @@ def test_scylla_sstable_query_temp_dir(cql, scylla_path, scylla_data_dir):
+        "sstable-dump-4d1cc6a4-6c13-11f0-a3c9",
+        "test-4d1cc6a46c1311f0a3c9"
+])
+def test_scylla_sstable_dump_data_from_snapshot(cql, test_keyspace, scylla_path, scylla_data_dir, tag):
```
It's impossible for a reader to guess what this test is testing and why, and what is special about the two tags you parametrized above. Please add a comment explaining what you are trying to reproduce here (please include a reference to the issue you're reproducing), and what kind of failure we saw in the past. In particular, in the code below, I see you just check that the "scylla sstables" command doesn't fail (you don't care what its actual output is) - so I assume it used to fail?

bhalevy replied:
> I see you just check that the "scylla sstables" command doesn't fail (you don't care what its actual output is) - so I assume it used to fail?

Yes, it is failing as of https://github.com/scylladb/scylladb/issues/25242 since the tool generates a wrong schema internally to query the sstable.

Will add json.loads validation on the tool's output to make that not only it succeeded, but it returned valid output

nyh commented on test/cqlpy/test_tools.py:
```diff
@@ -1737,6 +1737,22 @@ def test_scylla_sstable_query_null_data(cql, test_keyspace, scylla_path, scylla_
+        except subprocess.CalledProcessError as e:
+            pytest.fail(f"Failed to create snapshot: {e}\n{e.output.decode('utf-8')}")
+        args = [scylla_path, "sstable", "query", "--output-format", "json", "--scylla-yaml-file", f"{os.path.dirname(scylla_data_dir)}/conf/scylla.yaml", "--logger-log-level", "scylla-sstable=debug"]
+        args.extend([f"{os.path.dirname(sst)}/snapshots/{tag}/{os.path.basename(sst)}" for sst in sstables])
```
Please consider adding a "finally" that del_snapshot the snapshot you created.
Without that, what happens if you try to run this test twice (e.g., test/cqlpy/run --count=2 test_tools.py::test_scylla_sstable_dump_data_from_snapshot) - it will try to snapshot twice to the same name, does that work?

bhalevy replied:
This option doesn't seem to work anymore for `test/cqlpy/run`.  Tried with ./test.py --repeat=2 and there is not problem since each run uses a random keyspace and table names, but I'm ok with deleting the snapshot in any case.

bhalevy replied:
Weird, I'm running from the scylla dbuild docker container.
```
$ export SCYLLA=build/debug/scylla; test/cqlpy/run -v --count 2 test_tools.py::test_scylla_sstable_query_data_from_snapshot

Scylla is: /home/bhalevy/dev/scylla/build/debug/scylla.
Booting Scylla on 127.1.160.174 in /tmp/scylla-test-41134...
Boot successful (1.7 seconds).
Driver name ScyllaDB Python Driver, version 3.28.2
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --count
  inifile: /home/bhalevy/dev/scylla/test/pytest.ini
  rootdir: /home/bhalevy/dev/scylla/test
```

nyh replied:
Maybe dbuild doesn't have pytest-repeat installed - as I said it's needed for dtest (which can't run in dbuild as far as I remember), but once I installed it, I liked it and kept it, it's a generally useful plugin.

avikivity replied:
pytest-repeat wasn't added to the frozen toolchain yet.

nyh commented on test/cqlpy/test_tools.py:
```diff
@@ -1731,3 +1731,15 @@ def test_scylla_sstable_query_temp_dir(cql, scylla_path, scylla_data_dir):
         assert res.stderr.endswith(f"error running operation: std::filesystem::__cxx11::filesystem_error (error generic:20, filesystem error: temp_directory_path: Not a directory [{f.name}])\n")
+
+
+def test_scylla_sstable_query_null_data(cql, test_keyspace, scylla_path, scylla_data_dir):
```
Please add a comment explaining what this test reproduces (including mentioning the issue number). It's impossible to guess. I have no idea what "null data" refers to, or what was the error mode before it was fixed. Because I see your test doesn't actually check the content of the output - just that it didn't fail, right?

bhalevy commented:
In [7e40453](https://github.com/scylladb/scylladb/commit/7e40453ec5ef718ae85d785955e3006607810c3b):
- rebased on top of latest https://github.com/scylladb/scylladb/pull/25326
- Added comments about what the test tries to reproduce
- Used nodetool.take_snapshot instead of invoking the scylla nodetool binary
- Added nodetool.del_snapshot
- Added validation of json.loads(out) for the sstable query tool's output.

bhalevy commented:
> > Looks mostly good. It's still a mystery to me what "null data" refers to in the first patch, but other than that looks good.
> 
> The expected output contains null values:
> 
> ```
> [{"$pk0":3,"$pk1":3,"$ck0":0,"$ck1":0,"s":null,"v":0},{"$pk0":3,"$pk1":3,"$ck0":1,"$ck1":1,"s":null,"v":0},{"$pk0":3,"$pk1":3,"$ck0":2,"$ck1":2,"s":null,"v":0},{"$pk0":3,"$pk1":3,"$ck0":3,"$ck1":3,"s":null,"v":0},{"$pk0":3,"$pk1":3,"$ck0":5,"$ck1":5,"s":null,"v":0},{"$pk0":3,"$pk1":3,"$ck0":6,"$ck1":6,"s":null,"v":0},{"$pk0":3,"$pk1":3,"$ck0":8,"$ck1":8,"s":null,"v":0},{"$pk0":3,"$pk1":3,"$ck0":9,"$ck1":9,"s":null,"v":0},{"$pk0":0,"$pk1":0,"$ck0":2,"$ck1":2,"s":null,"v":0},{"$pk0":0,"$pk1":0,"$ck0":3,"$ck1":3,"s":null,"v":0},{"$pk0":0,"$pk1":0,"$ck0":5,"$ck1":5,"s":null,"v":0},{"$pk0":0,"$pk1":0,"$ck0":6,"$ck1":6,"s":null,"v":0},{"$pk0":0,"$pk1":0,"$ck0":8,"$ck1":8,"s":null,"v":0},{"$pk0":0,"$pk1":0,"$ck0":9,"$ck1":9,"s":null,"v":0},{"$pk0":5,"$pk1":5,"$ck0":0,"$ck1":0,"s":10,"v":0},{"$pk0":5,"$pk1":5,"$ck0":1,"$ck1":1,"s":10,"v":0},{"$pk0":5,"$pk1":5,"$ck0":3,"$ck1":3,"s":10,"v":0},{"$pk0":5,"$pk1":5,"$ck0":4,"$ck1":4,"s":10,"v":0},{"$pk0":5,"$pk1":5,"$ck0":7,"$ck1":7,"s":10,"v":0},{"$pk0":5,"$pk1":5,"$ck0":8,"$ck1":8,"s":10,"v":0},{"$pk0":5,"$pk1":5,"$ck0":9,"$ck1":9,"s":10,"v":0},{"$pk0":2,"$pk1":2,"$ck0":2,"$ck1":2,"s":null,"v":0},{"$pk0":2,"$pk1":2,"$ck0":3,"$ck1":3,"s":null,"v":0},{"$pk0":2,"$pk1":2,"$ck0":4,"$ck1":4,"s":null,"v":0},{"$pk0":2,"$pk1":2,"$ck0":7,"$ck1":7,"s":null,"v":0},{"$pk0":2,"$pk1":2,"$ck0":9,"$ck1":9,"s":null,"v":0},{"$pk0":7,"$pk1":7,"$ck0":3,"$ck1":3,"s":null,"v":0},{"$pk0":7,"$pk1":7,"$ck0":4,"$ck1":4,"s":null,"v":0},{"$pk0":7,"$pk1":7,"$ck0":8,"$ck1":8,"s":null,"v":0},{"$pk0":7,"$pk1":7,"$ck0":9,"$ck1":9,"s":null,"v":0},{"$pk0":9,"$pk1":9,"$ck0":0,"$ck1":0,"s":null,"v":0},{"$pk0":9,"$pk1":9,"$ck0":1,"$ck1":1,"s":null,"v":0},{"$pk0":9,"$pk1":9,"$ck0":2,"$ck1":2,"s":null,"v":0},{"$pk0":9,"$pk1":9,"$ck0":5,"$ck1":5,"s":null,"v":0},{"$pk0":9,"$pk1":9,"$ck0":6,"$ck1":6,"s":null,"v":0},{"$pk0":9,"$pk1":9,"$ck0":8,"$ck1":8,"s":null,"v":0},{"$pk0":8,"$pk1":8,"$ck0":2,"$ck1":2,"s":null,"v":0},{"$pk0":8,"$pk1":8,"$ck0":4,"$ck1":4,"s":null,"v":0},{"$pk0":8,"$pk1":8,"$ck0":6,"$ck1":6,"s":null,"v":0},{"$pk0":8,"$pk1":8,"$ck0":8,"$ck1":8,"s":null,"v":0},{"$pk0":8,"$pk1":8,"$ck0":9,"$ck1":9,"s":null,"v":0},{"$pk0":6,"$pk1":6,"$ck0":1,"$ck1":1,"s":null,"v":0},{"$pk0":6,"$pk1":6,"$ck0":2,"$ck1":2,"s":null,"v":0},{"$pk0":6,"$pk1":6,"$ck0":4,"$ck1":4,"s":null,"v":0},{"$pk0":6,"$pk1":6,"$ck0":5,"$ck1":5,"s":null,"v":0},{"$pk0":6,"$pk1":6,"$ck0":6,"$ck1":6,"s":null,"v":0},{"$pk0":6,"$pk1":6,"$ck0":7,"$ck1":7,"s":null,"v":0},{"$pk0":6,"$pk1":6,"$ck0":8,"$ck1":8,"s":null,"v":0},{"$pk0":1,"$pk1":1,"$ck0":0,"$ck1":0,"s":null,"v":0},{"$pk0":1,"$pk1":1,"$ck0":2,"$ck1":2,"s":null,"v":0},{"$pk0":1,"$pk1":1,"$ck0":3,"$ck1":3,"s":null,"v":0},{"$pk0":1,"$pk1":1,"$ck0":4,"$ck1":4,"s":null,"v":0},{"$pk0":1,"$pk1":1,"$ck0":5,"$ck1":5,"s":null,"v":0},{"$pk0":1,"$pk1":1,"$ck0":6,"$ck1":6,"s":null,"v":0},{"$pk0":1,"$pk1":1,"$ck0":7,"$ck1":7,"s":null,"v":0},{"$pk0":1,"$pk1":1,"$ck0":9,"$ck1":9,"s":null,"v":0},{"$pk0":4,"$pk1":4,"$ck0":2,"$ck1":2,"s":null,"v":0},{"$pk0":4,"$pk1":4,"$ck0":5,"$ck1":5,"s":null,"v":0},{"$pk0":4,"$pk1":4,"$ck0":6,"$ck1":6,"s":null,"v":0},{"$pk0":4,"$pk1":4,"$ck0":7,"$ck1":7,"s":null,"v":0},{"$pk0":4,"$pk1":4,"$ck0":8,"$ck1":8,"s":null,"v":0},{"$pk0":4,"$pk1":4,"$ck0":9,"$ck1":9,"s":null,"v":0}]
> ```
> 
> `simple_clustering_table` set `s` only fro a single key (5) and it is null on others. I can add a test that null is present in the output to make sure it was processed properly.

This actually refers more to https://github.com/scylladb/scylladb/pull/25326#issuecomment-3157891251 not this PR.

bhalevy commented:
[0a7d72b](https://github.com/scylladb/scylladb/commit/0a7d72becaf0cf282b8e9804af8776b8b0c74c13):
- rebased on top of 5e5e63af10
